src/netlink_namespace.c

```diff
@@ -0,0 +1,817 @@
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; only version 2 of the License is applicable.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ *
+ * Authors:
+ *   SRE Team at Medallia Inc. <eng-sre at medallia.com>
+ *   Florian octo Forster <octo at collectd.org>
+ *   Sebastian Harl <sh at tokkee.org>
+ *   Andreas Henriksson <andreas at fatal.se>
+ *   Marc Fournier <marc.fournier at camptocamp.com>
+ **/
+
+#define _GNU_SOURCE
+#include <sched.h>
+
+#include "collectd.h"
+
+#include "common.h"
+#include "plugin.h"
+
+#include <asm/types.h>
+
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+#if HAVE_LINUX_GEN_STATS_H
+#include <linux/gen_stats.h>
+#endif
+#if HAVE_LINUX_PKT_SCHED_H
+#include <linux/pkt_sched.h>
+#endif
+
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include <unistd.h>
+
+#include <libmnl/libmnl.h>
+
+struct ir_link_stats_storage_s {
+
+  uint64_t rx_packets;
+  uint64_t tx_packets;
+  uint64_t rx_bytes;
+  uint64_t tx_bytes;
+  uint64_t rx_errors;
+  uint64_t tx_errors;
+
+  uint64_t rx_dropped;
+  uint64_t tx_dropped;
+  uint64_t multicast;
+  uint64_t collisions;
+
+  uint64_t rx_length_errors;
+  uint64_t rx_over_errors;
+  uint64_t rx_crc_errors;
+  uint64_t rx_frame_errors;
+  uint64_t rx_fifo_errors;
+  uint64_t rx_missed_errors;
+
+  uint64_t tx_aborted_errors;
+  uint64_t tx_carrier_errors;
+  uint64_t tx_fifo_errors;
+  uint64_t tx_heartbeat_errors;
+  uint64_t tx_window_errors;
+};
+
+union ir_link_stats_u {
+  struct rtnl_link_stats *stats32;
+#ifdef HAVE_RTNL_LINK_STATS64
+  struct rtnl_link_stats64 *stats64;
+#endif
+};
+
+typedef struct ir_ignorelist_s {
+  char *device;
+  char *type;
+  char *inst;
+  struct ir_ignorelist_s *next;
+} ir_ignorelist_t;
+
+struct qos_stats {
+  struct gnet_stats_basic *bs;
+  struct gnet_stats_queue *qs;
+};
+
+static int ir_ignorelist_invert = 1;
+static ir_ignorelist_t *ir_ignorelist_head = NULL;
+
+static struct mnl_socket *nl;
+
+static char **iflist = NULL;
+static size_t iflist_len = 0;
+
+static const char *config_keys[] = {"Interface", "VerboseInterface",
+                                    "QDisc",     "Class",
+                                    "Filter",    "IgnoreSelected",
+                                    "NSName"};
+static int config_keys_num = STATIC_ARRAY_SIZE(config_keys);
+
+/* Will store namespace name here */
+static char **ns_name = NULL;
```

Comment:
The history is basically that I started looking for ways to have multiple namespace support, but then I went less ambitious.